### PR TITLE
Support mbedtls crypto library

### PR DESCRIPTION
**Pull request by edison-ai**
Move  #2345 here.

### CODE REVIEW

**Comment by jenswi-linaro** [comment]
There seems to be some confusion here. Only the commits:
"libmbedtls: GCM and CCM enhance for integrate", "libmbedtls: fix no CRT issue" and "libmbedtls: add interfaces in mbedtls for context memory operation" should first be merged into `import/mbedtls-2.6.1`.

The purpose of the `import/mbedtls-2.6.1` is to record all changes to the imports source code of mbedtls-2.6.1. This will then be used when we import a newer version of mbedtls.

Since `import/mbedtls-2.6.1` happens to be quite close to `master` for the moment it's still meaningful to review the changes here. It's just that before anything can be merged we'll need a new PR against the `master` branch. I'll continue the review now.

File: lib/libmbedtls/core/sub.mk [hunk 1]
```diff
@@ -1,3 +1,20 @@
 cflags-lib-$(CFG_CRYPTO_SIZE_OPTIMIZATION) += -Os
 
 srcs-y += tee_lmd_provider.c
+
+ifneq (,$(filter y, $(CFG_CRYPTO_DSA)))
```
Comment (jenswi-linaro) [inline lib/libmbedtls/core/sub.mk]: Why isn't `ifeq ($(CFG_CRYPTO_DSA),y)` used?

Comment (jenswi-linaro) [reply lib/libmbedtls/core/sub.mk]: Please ignore, see that now in later commits.

File: lib/libmbedtls/core/tee_lmd_provider.c [hunk 2]
```diff
@@ -1931,11 +1990,13 @@ static TEE_Result cipher_get_ctx_size(uint32_t algo, size_t *size)
 #endif
 #if defined(CFG_CRYPTO_XTS)
 	case TEE_ALG_AES_XTS:
-		return TEE_ERROR_NOT_SUPPORTED;
+		*size = sizeof(struct tee_symmetric_cts);
```
Comment (jenswi-linaro) [inline lib/libmbedtls/core/tee_lmd_provider.c]: tee_symmetric_xts

File: lib/libmbedtls/core/tee_lmd_provider.c [hunk 3]
```diff
@@ -1931,11 +1990,13 @@ static TEE_Result cipher_get_ctx_size(uint32_t algo, size_t *size)
...
+		break;
 #endif
 #if defined(CFG_CRYPTO_CTS)
 	case TEE_ALG_AES_CTS:
-		return TEE_ERROR_NOT_SUPPORTED;
+		*size = sizeof(struct tee_symmetric_xts);
```
Comment (jenswi-linaro) [inline lib/libmbedtls/core/tee_lmd_provider.c]: tee_symmetric_cts

File: lib/libmbedtls/core/tee_lmd_provider.c [hunk 4]
```diff
@@ -2016,8 +2077,10 @@ TEE_Result crypto_cipher_alloc_ctx(void **ctx_ret, uint32_t algo)
 		return TEE_ERROR_NOT_SUPPORTED;
 	}
 
-	if (!cipher_info)
+	if (algo != TEE_ALG_AES_XTS && algo != TEE_ALG_AES_CTS &&
+	    !cipher_info) {
```
Comment (jenswi-linaro) [inline lib/libmbedtls/core/tee_lmd_provider.c]: Minor: the braces aren't needed.

File: lib/libmbedtls/core/tee_lmd_provider.c [hunk 5]
```diff
@@ -2049,13 +2116,26 @@ void crypto_cipher_free_ctx(void *ctx, uint32_t algo __maybe_unused)
 	 * could never have succeded above.
 	 */
 	assert(!cipher_get_ctx_size(algo, &ctx_size));
-	mbedtls_cipher_free(ctx);
+#if defined(CFG_CRYPTO_XTS) || defined(CFG_CRYPTO_CTS)
```
Comment (jenswi-linaro) [inline lib/libmbedtls/core/tee_lmd_provider.c]: Unless `CFG_CRYPTO_XTS` or `CFG_CRYPTO_CTS` is defined we'll potentially leak memory.
I see no harm in always having the `if (...)` below

File: lib/libmbedtls/core/tee_lmd_provider.c [hunk 6]
```diff
@@ -2049,13 +2116,26 @@ void crypto_cipher_free_ctx(void *ctx, uint32_t algo __maybe_unused)
...
 				uint32_t algo __unused)
 {
+	TEE_Result res __maybe_unused;
+	size_t ctx_size __maybe_unused;
+
+#if defined(CFG_CRYPTO_XTS) || defined(CFG_CRYPTO_CTS)
+	if (algo == TEE_ALG_AES_XTS || algo == TEE_ALG_AES_CTS) {
+		res = cipher_get_ctx_size(algo, &ctx_size);
+		assert(!res);
+		memcpy(dst_ctx, src_ctx, ctx_size);
```
Comment (jenswi-linaro) [inline lib/libmbedtls/core/tee_lmd_provider.c]: I think a `return;` is needed after this line.

File: lib/libmbedtls/core/tee_lmd_provider.c [hunk 7]
```diff
@@ -2071,6 +2151,58 @@ TEE_Result crypto_cipher_init(void *ctx, uint32_t algo,
 	const mbedtls_cipher_info_t *cipher_info = NULL;
 	int lmd_res;
 	int lmd_mode;
+#if defined(CFG_CRYPTO_CTS)
+	struct tee_symmetric_cts *cts;
```
Comment (jenswi-linaro) [inline lib/libmbedtls/core/tee_lmd_provider.c]: These helper variables are only used in their if blocks below so please move the declarations.

File: lib/libmbedtls/core/tee_lmd_provider.c [hunk 8]
```diff
@@ -2071,6 +2151,58 @@ TEE_Result crypto_cipher_init(void *ctx, uint32_t algo,
...
+#if defined(CFG_CRYPTO_XTS)
+	struct tee_symmetric_xts *xts;
+#endif
+	TEE_Result res __maybe_unused;
+	int ltc_res __maybe_unused;
+	int ltc_cipherindex __maybe_unused;
+
+#if defined(CFG_CRYPTO_XTS)
+	if (algo == TEE_ALG_AES_XTS) {
+		res = tee_algo_to_ltc_cipherindex(algo, &ltc_cipherindex);
```
Comment (jenswi-linaro) [inline lib/libmbedtls/core/tee_lmd_provider.c]: Could move these three lines into the other `if (algo == TEE_ALG_AES_XTS)` block below.

File: lib/libmbedtls/core/tee_lmd_provider.c [hunk 9]
```diff
@@ -2127,6 +2259,13 @@ TEE_Result crypto_cipher_update(void *ctx, uint32_t algo,
 	int lmd_res;
 	size_t olen;
 	size_t finish_olen;
+#if defined(CFG_CRYPTO_CTS)
+	struct tee_symmetric_cts *cts;
```
Comment (jenswi-linaro) [inline lib/libmbedtls/core/tee_lmd_provider.c]: I'd prefer if the `#if` was skipped and the `__maybe_unused` tag was used instead, here and the two other locals below.

File: lib/libmbedtls/core/tee_lmd_provider.c [hunk 10]
```diff
@@ -2383,6 +2383,46 @@ void crypto_cipher_final(void *ctx __unused, uint32_t algo __unused)
  * Message Authentication Code functions
  *****************************************************************************/
 #if defined(_CFG_CRYPTO_WITH_MAC)
+
+#if defined(CFG_CRYPTO_CBC_MAC)
+/*
+ * CBC-MAC is not implemented in Libtomcrypt
```
Comment (jenswi-linaro) [inline lib/libmbedtls/core/tee_lmd_provider.c]: `s/Libtomcrypt/mbedtls/`

File: lib/libmbedtls/core/tee_lmd_provider.c [hunk 11]
```diff
@@ -2629,7 +2683,31 @@ TEE_Result crypto_mac_init(void *ctx, uint32_t algo, const uint8_t *key,
 	case TEE_ALG_DES_CBC_MAC_PKCS5:
 	case TEE_ALG_DES3_CBC_MAC_NOPAD:
 	case TEE_ALG_DES3_CBC_MAC_PKCS5:
-		return TEE_ERROR_NOT_SUPPORTED;
+		cbc = (struct cbc_state *)ctx;
+
+		res = tee_algo_to_ltc_cipherindex(algo, &ltc_cipherindex);
```
Comment (jenswi-linaro) [inline lib/libmbedtls/core/tee_lmd_provider.c]: How come the mbedtls implementation of AES/DES/DES3 isn't used instead?

**Comment by jenswi-linaro** [comment]
For "libmbedtls: Support DSA algorithm" please apply:
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

**Comment by jenswi-linaro** [comment]
I'm done with the review for the moment.

File: lib/libmbedtls/core/tee_lmd_provider.c (hunk 2 above)
Comment (edison-ai) [reply lib/libmbedtls/core/tee_lmd_provider.c]: Yes, sorry for mistake.

File: lib/libmbedtls/core/tee_lmd_provider.c (hunk 3 above)
Comment (edison-ai) [reply lib/libmbedtls/core/tee_lmd_provider.c]: Yes, will fix it.

File: lib/libmbedtls/core/tee_lmd_provider.c (hunk 4 above)
Comment (edison-ai) [reply lib/libmbedtls/core/tee_lmd_provider.c]: Yes, will fix it.

File: lib/libmbedtls/core/tee_lmd_provider.c (hunk 5 above)
Comment (edison-ai) [reply lib/libmbedtls/core/tee_lmd_provider.c]: Yes, it is a mistake. Should not add them here.

File: lib/libmbedtls/core/tee_lmd_provider.c (hunk 6 above)
Comment (edison-ai) [reply lib/libmbedtls/core/tee_lmd_provider.c]: OK, will add it.

File: lib/libmbedtls/core/tee_lmd_provider.c (hunk 7 above)
Comment (edison-ai) [reply lib/libmbedtls/core/tee_lmd_provider.c]: OK.

File: lib/libmbedtls/core/tee_lmd_provider.c (hunk 8 above)
Comment (edison-ai) [reply lib/libmbedtls/core/tee_lmd_provider.c]: OK

File: lib/libmbedtls/core/tee_lmd_provider.c (hunk 9 above)
Comment (edison-ai) [reply lib/libmbedtls/core/tee_lmd_provider.c]: Sorry for just copy code to here directly. I will move the declare to the block where use it.

File: lib/libmbedtls/core/tee_lmd_provider.c (hunk 10 above)
Comment (edison-ai) [reply lib/libmbedtls/core/tee_lmd_provider.c]: Yes, I will fix it.

File: lib/libmbedtls/core/tee_lmd_provider.c (hunk 11 above)
Comment (edison-ai) [reply lib/libmbedtls/core/tee_lmd_provider.c]: Sorry not notice it before. I will refine it by using mbedtls.

**Comment by edison-ai** [comment]
For "libmbedtls: GCM and CCM enhance for integrate", "libmbedtls: fix no CRT issue" and "libmbedtls: add interfaces in mbedtls for context memory operation", I will drop them in this PR. Because they had been in #2439.

**Comment by jenswi-linaro** [comment]
We've advanced far enough in this PR that it would make sense to start merging that #2439.

**Comment by jenswi-linaro** [comment]
Replaced by the now merged https://github.com/OP-TEE/optee_os/pull/2904

**Pull request closed without merging**